### ISSUE TITLE
Dataframe write command very slow when database is already populated

### ISSUE THREAD

**Issue by holypriest** [RedisLabs/spark-redis#131]
I noticed that the writing times are way different when the Redis database is empty and when there are keys inside, already.

I wrote a dataframe with ~ 2 million rows (PySpark), and it happened in a blazing-fast manner when the DB was empty (~ 3 minutes). Then, I tried to write the same dataframe (changing the column-key, to avoid overwrites) and it took ~ 50 minutes. After executing FLUSHDB I got the same 3 minutes write as before.

Is this behavior expected somehow? Is there a way to overcome that issue?

**Comment by fe2s** [comment]
Hi,

Do you mean that you change "key.column" option or "table" option? 
Do you specify any SaveMode?

Ideally, please share you code.

**Comment by holypriest** [comment]
Hi, @fe2s. That's a sample of my code, that takes care of the writing part:

```
df.write \
  .format("org.apache.spark.sql.redis") \
  .option("table", foo) \
  .option("key.column", "id") \
  .option("ttl", time_to_expire) \
  .save()
```

I changed the `table` option (assigning a new value to the `foo` variable, same size in bytes). For the rest, I kept the same values.

I tried to set the SaveMode to `overwrite`, but I got the same writing times.

**Comment by fe2s** [comment]
Could you please try `.mode('append')`?

**Comment by holypriest** [comment]
Sure, I will come back with the results.

**Comment by holypriest** [comment]
Yes, apparently it _does_ work! But why? Is there something happening on the Redis size, skipping key checks or something?

**Comment by fe2s** [comment]
With the default `SaveMode (ErrorIfExists)` spark checks if the dataframe already exists and returns an error in that case.
With `.mode('append')` it skips the check and just writes the data.
The check is implemented by scanning keys with the `table:*` pattern. It looks like that for some reason it takes a lot of time (almost 50 mins) in your case.

I tried to reproduce it on my laptop (one local Redis node, one local Spark worker)

```python
import uuid
import time
import sys
from pyspark.sql import SparkSession
from pyspark.sql import SQLContext

if __name__ == "__main__":
    spark = SparkSession \
        .builder \
        .appName("df") \
        .getOrCreate()

    sc = spark.sparkContext
    sqlContext = SQLContext(sc)

    def createDf(num):
        partitions = 10
        rdd = sc.parallelize(range(1, partitions), partitions).mapPartitions(
            lambda x: [(uuid.uuid4().hex, i) for i in range(num / partitions)])
        return sqlContext.createDataFrame(rdd, ['name', 'age'])

    num = 3000000
    df = createDf(num)

    start = time.time()
    table = sys.argv[1]
    df.write.format("org.apache.spark.sql.redis").option("table", table).option("key.column", "name").save()
    end = time.time()
    print("time taken to save df: " + str(end - start))

    spark.stop()
```

With the first run (empty Redis):
* it takes 0.182 sec to check if dataframe already exists
* it takes 33 sec to save dataframe

With the second run (another table name):
* it takes 8.787 sec to check if dataframe already exists
* it takes 39 sec to save dataframe

Would you be able to do the following:
* share your spark-redis version
* share your Spark cluster details (number of workers, allocated memory)
* share driver and executor logs
* run my example on your cluster and see how long it takes

The optimization we can do is to just check the existence of the Redis key that holds dataframe schema, this will be an indicator if dataframe exists. The downside of this is that we will not track Redis keys that were written by non-spark applications.
